bson/src/main/org/bson/codecs/pojo/MapPropertyCodecProvider.java

```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright 2017 MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.bson.codecs.pojo;
+
+import org.bson.BsonReader;
+import org.bson.BsonType;
+import org.bson.BsonWriter;
+import org.bson.codecs.Codec;
+import org.bson.codecs.DecoderContext;
+import org.bson.codecs.EncoderContext;
+import org.bson.codecs.configuration.CodecConfigurationException;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Map.Entry;
+
+import static java.lang.String.format;
+
+final class MapPropertyCodecProvider implements PropertyCodecProvider {
+
+    @SuppressWarnings({"rawtypes", "unchecked"})
+    @Override
+    public <T> Codec<T> get(final TypeWithTypeParameters<T> type, final PropertyCodecRegistry registry) {
+        if (Map.class.isAssignableFrom(type.getType()) && type.getTypeParameters().size() == 2) {
+            Class<?> keyType = type.getTypeParameters().get(0).getType();
+            if (!keyType.equals(String.class)) {
+                throw new CodecConfigurationException(format("Invalid Map type. Maps MUST have string keys, found %s instead.", keyType));
```

Comment:
Previously, TypeData would have thrown an IllegalArgument Exception for a Map without a String key.